File: server/TWS Foundation/CSM Foundation/Database/Entity/Models/Output/SetViewOutput.cs
namespace CSM_Foundation.Database.Entity.Models.Output;

/// <summary>
///     Stores a result <see cref="SetViewOutput{TMigrationSet}"/> 
///     after a build operation determining paging, filtering and ordering.
/// </summary>
/// <typeparam name="TEntity">
///     <typeparamref name="TEntity"/> that this view handles results.
/// </typeparam>
public class SetViewOutput<TEntity>
    where TEntity : IEntity {

    TEntity[] _Records = [];
    /// <summary>
    ///     The collection of items gathered.
    /// </summary>
    public required TEntity[] Records {
        get => _Records;
        init {
            _Records = value;
            Length = value.Length;
        }
    }
    /// <summary>
    ///     The available pages.
    /// </summary>
    public required int Pages { get; init; }

    /// <summary>
    ///     The current page.
    /// </summary>
    public required int Page { get; init; }

    /// <summary>
    ///     Indicates the timemark when was created.
    /// </summary>
    public DateTime Timestamp { get; init; } = DateTime.UtcNow;

    /// <summary>
    ///     Indicates the quantity of records that this result contains.
    /// </summary>
    public int Length { get; init; }

    /// <summary>
    ///     Count of total records that currently exist at the live database
    /// </summary>
    public required int Count { get; init; }
}

File: client/tws_foundation_client/lib/src/models/outputs/set_view_output.dart
import 'package:csm_client/csm_client.dart';

/// Defines the mirror class to store a [SetViewOutput] result,
/// describing the behavior of a paged table returning data for the required size of pages and page.
///
///
/// [TSet] : The type of live database mirror set from to build the [View].
final class SetViewOutput<TSet extends EntityB<TSet>> implements EncodableI, DecodableI {
  /// Records resolved.
  List<TSet> records;

  /// Total amount of pages available.
  int pages;

  /// The current page resulted.
  int page;

  /// Timemark where this view resolution was created.
  DateTime creation;

  /// The quantity of records resolved.
  int length;

  /// The total quantity of available records at the data storage.
  int count;

  /// Creates a new [SetViewOutput] object.
  SetViewOutput(this.records, this.page, this.creation, this.pages, this.length, this.count);

  @override
  DataMap encode() {
    return <String, dynamic>{
      'records': records.map((TSet e) => e.encode()).toList(),
      'pages': pages,
      'page': page,
      'creation': creation.toString(),
      'length': length,
      'count': count,
    };
  }
  
  @override
  void decode(DataMap encode) {
    records = encode.getList('records');
    pages = encode.get('pages');
    page = encode.get('page');
    creation = encode.get('creation');
    length = encode.get('lenght');
    count = encode.get('count');
  }
}
